### PR TITLE
Add improve-analysis-1 to master

### PR DESCRIPTION
**Pull request by Aichnerc**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?

**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
- Currently, you can not create `zombies` out of `cats`.
- Currently, the image of the user does not load properly.
- Currently, the radar chart for the analysis looks not that great.
- Currently, the dashboard is quite clumpy.

**What is the new behavior (if this is a feature change)?**
- You can now promote `cats` to `zombies`.
- The image has been added to Firestore and is now loaded, including `lazy loading` the image.
- The analysis radar chart has been improved and now looks like its counterpart on the website. Ref: [www.aichner-christian.com](www.aichner-christian.com)
- The dashboard has been seperated into multiple phases to clear up the view.

**Other information**:
- The version has been updated to v1.2.0.

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -37,7 +33,7 @@ class App extends React.Component {
 
 export default App;
 
-/** 
+/**
  * SPDX-License-Identifier: (EUPL-1.2)
  * Copyright © 2019 Werbeagentur Christian Aichner
```
Comment (LucaAllmaier) [inline src/App.js]: missing -2020

File: src/Routes.js [hunk 2]
```diff
@@ -49,7 +45,7 @@ class Routes extends React.Component {
 
 export default Routes;
 
-/** 
+/**
  * SPDX-License-Identifier: (EUPL-1.2)
  * Copyright © 2019 Werbeagentur Christian Aichner
```
Comment (LucaAllmaier) [inline src/Routes.js]: missing -2020

File: src/components/organisms/ZombieList/index.jsx [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+            </>
+          </div>
+        </>
+      );
+    }
+  }
+}
+
+const mapStateToProps = (state) => {
+  console.log(state);
```
Comment (LucaAllmaier) [inline src/components/organisms/ZombieList/index.jsx]: is this still needed?

**Review by LucaAllmaier — COMMENTED** [review]
Looks good.

**Review by sweggerat — APPROVED** [review]
looks good to me

**Review by LucaAllmaier — APPROVED** [review]
Looks good to merge.

File: src/components/molecules/charts/ResultChart/index.jsx [hunk 4]
```diff
@@ -77,10 +77,22 @@ class ResultChart extends React.Component {
...
+          orderedData[key] = unsortedData[key];
+        });
+
       // Get value and label
-      Object.keys(this.props.data).map((keyName, i) => {
-        labels.push(this.props.data[keyName].name);
-        data.push(this.props.data[keyName].value);
+      console.log(Object.keys(orderedData));
+      Object.keys(orderedData).map((keyName, i) => {
+        console.log(keyName);
```
Comment (schettn) [inline src/components/molecules/charts/ResultChart/index.jsx]: Add new line after the console log or remove the log.

File: src/components/molecules/charts/ResultChart/index.jsx [hunk 5]
```diff
@@ -77,10 +77,22 @@ class ResultChart extends React.Component {
       labels = [];
       data = [];
 
+      // Initialize data
+      const unsortedData = this.props.data;
+      // Sort object
+      const orderedData = {};
```
Comment (schettn) [inline src/components/molecules/charts/ResultChart/index.jsx:83]: add newline

File: src/components/molecules/modals/AboutUs/index.jsx [hunk 6]
```diff
@@ -4,90 +4,97 @@ import React from "react";
...
   state = {
-    modal11: false
-  }
+    modal11: false,
+  };
 
-  toggle = nr => () => {
-    let modalNumber = "modal" + nr
+  toggle = (nr) => () => {
+    let modalNumber = "modal" + nr;
```
Comment (schettn) [inline src/components/molecules/modals/AboutUs/index.jsx:25]: add newline

File: src/components/organisms/CatList/index.jsx [hunk 7]
```diff
@@ -105,9 +104,16 @@ class CatList extends React.Component {
     }
   };
 
+  updateCat = (cat) => {
+    cat = {
+      ...cat,
+      password: Math.random().toString(36).slice(-8),
+    };
```
Comment (schettn) [inline src/components/organisms/CatList/index.jsx:111]: add newline

File: src/components/molecules/modals/AboutUs/index.jsx [hunk 8]
```diff
@@ -4,90 +4,97 @@ import React from "react";
...
 class ModalPage extends React.Component {
   state = {
-    modal11: false
-  }
+    modal11: false,
+  };
 
-  toggle = nr => () => {
-    let modalNumber = "modal" + nr
+  toggle = (nr) => () => {
```
Comment (schettn) [inline src/components/molecules/modals/AboutUs/index.jsx]: Is this useful? I would suggest changing it to:
foo = (param) => {}

File: src/components/organisms/ZombieList/index.jsx [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+        modalUser: undefined,
+      });
+    }
+  };
+
+  updateCat = (cat) => {
+    cat = {
+      ...cat,
+      password: Math.random().toString(36).slice(-8),
+    };
```
Comment (schettn) [inline src/components/organisms/ZombieList/index.jsx]: add newline

File: src/components/organisms/ZombieList/index.jsx [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    this.setState({
+      data: {
+        ...this.state.data,
+        rows: this.getUserList(zombies),
+      },
+    });
+  };
+
+  render() {
+    const { auth, profile, zombies } = this.props;
```
Comment (schettn) [inline src/components/organisms/ZombieList/index.jsx:161]: Add newline or remove the console log below.

File: src/components/pages/MessagePage/index.jsx [hunk 11]
```diff
@@ -21,7 +21,7 @@ import "./messagepage.scss";
 class MessagePage extends React.Component {
   render() {
     const { auth, profile } = this.props;
```
Comment (schettn) [inline src/components/pages/MessagePage/index.jsx:23]: add newline or remove the console log

File: src/components/pages/ProfilePage/index.jsx [hunk 12]
```diff
@@ -61,11 +61,11 @@ class ProfilePage extends React.Component {
     // Store selected greeting
     let selected = null;
```
Comment (schettn) [inline src/components/pages/ProfilePage/index.jsx:62]: add newline or remove the console log below

File: src/components/pages/ProfilePage/index.jsx [hunk 13]
```diff
@@ -61,11 +61,11 @@ class ProfilePage extends React.Component {
...
     console.log(curHr);
```
Comment (schettn) [inline src/components/pages/ProfilePage/index.jsx]: add newline

File: src/config/fbInit.js [hunk 14]
```diff
@@ -1,25 +1,25 @@
...
+import "firebase/firestore";
+import "firebase/auth";
+import "firebase/analytics";
 
 // Import secret credentials
-import config from "./fbConfig.js"
+import config from "./fbConfig.js";
 
 // Init Firebase
 firebase.initializeApp(config);
```
Comment (schettn) [inline src/config/fbInit.js:11]: add newline

File: src/store/actions/postActions.js [hunk 15]
```diff
@@ -1,184 +1,226 @@
...
+      .collection("posts")
+      .doc(uniqueID)
+      .set({
+        ...newPost,
+        likes: [],
+        skin: newPost.skin ? newPost.skin : null,
+        visible: true,
+      })
+      .then(() => {
+        dispatch({ type: "CREATION_SUCCESS", newPost });
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 16]
```diff
@@ -1,184 +1,226 @@
...
+      .collection("posts")
+      .doc(id)
+      .set(
+        {
+          likes: localLikes,
+        },
+        { merge: true }
+      )
+      .then(() => {
+        dispatch({ type: "LIKE_SUCCESS", id });
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 17]
```diff
@@ -1,184 +1,226 @@
...
+
+    // Create post
+    firestore
+      .collection("posts")
+      .doc(id)
+      .update({
+        likes: likes,
+      })
+      .then(() => {
+        dispatch({ type: "UNLIKE_SUCCESS", id });
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 18]
```diff
@@ -1,184 +1,226 @@
...
+        .collection("posts")
+        .doc(post.id)
+        .set(
+          {
+            visible: false,
+          },
+          { merge: true }
+        )
+        .then(() => {
+          dispatch({ type: "REMOVE_SUCCESS", id: post.id });
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 19]
```diff
@@ -1,184 +1,226 @@
...
-            }).catch((err) => {
-                dispatch({ type: "REMOVE_ERROR", err });
-            })
-        } else {
-            dispatch({ type: "REMOVE_ERROR", err: "Not authorized." });
-        }
+export const loadPosts = (amount) => {
+  return (dispatch, getState, { getFirebase, getFirestore }) => {
+    const firebase = getFirebase();
+    const firestore = getFirestore();
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 20]
```diff
@@ -1,184 +1,226 @@
...
-            dispatch({ type: "LOAD_SUCCESS", results });
-        })
-        .catch((err) => {
-            dispatch({ type: "LOAD_ERROR", err });
+    posts
+      .orderBy("timestamp", "desc")
+      .limit(amount)
+      .get()
+      .then((querySnapshot) => {
+        let results = [];
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 21]
```diff
@@ -1,184 +1,226 @@
...
+        querySnapshot.forEach(function (doc) {
+          let data = doc.data();
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 22]
```diff
@@ -1,184 +1,226 @@
...
+  return (dispatch, getState, { getFirebase, getFirestore }) => {
+    const firebase = getFirebase();
+    const firestore = getFirestore();
+
+    let localComments = previousComments;
+    if (localComments) {
+      if (Array.isArray(localComments)) {
+        localComments.push(comment);
+      } else {
+        localComments = [];
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 23]
```diff
@@ -1,184 +1,226 @@
...
+        localComments.push(comment);
+      }
+    } else {
+      localComments = [];
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 24]
```diff
@@ -1,184 +1,226 @@
...
+      .collection("posts")
+      .doc(postId)
+      .set(
+        {
+          comments: localComments,
+        },
+        { merge: true }
+      )
+      .then(() => {
+        dispatch({ type: "COMMENT_SUCCESS", postId });
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 25]
```diff
@@ -1,184 +1,226 @@
...
-                results.push({id: doc.id, data});
-            });
-            dispatch({ type: "LOAD_SUCCESS", results });
-        })
-        .catch((err) => {
-            dispatch({ type: "LOAD_ERROR", err });
-        });
+  return (dispatch, getState, { getFirebase, getFirestore }) => {
+    const firebase = getFirebase();
+    const firestore = getFirestore();
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 26]
```diff
@@ -1,184 +1,226 @@
...
     }
-}
 
-/** 
+    posts
+      .orderBy("timestamp", "desc")
+      .limit(amount)
+      .get()
+      .then((querySnapshot) => {
+        let results = [];
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/postActions.js [hunk 27]
```diff
@@ -1,184 +1,226 @@
...
+        querySnapshot.forEach(function (doc) {
+          let data = doc.data();
```
Comment (schettn) [inline src/store/actions/postActions.js]: add newline

File: src/store/actions/userActions.js [hunk 28]
```diff
@@ -1,37 +1,40 @@
...
-        });
-    }
-}
+    firestore
+      .collection("users")
+      .doc(uid)
+      .get()
+      .then((doc) => {
+        if (!doc.exists) {
+          console.log("User no longer exists!");
```
Comment (schettn) [inline src/store/actions/userActions.js]: add newline or remove the console log

File: src/store/actions/userActions.js [hunk 29]
```diff
@@ -1,37 +1,40 @@
...
+          dispatch({ type: "USERLOAD_ERROR", err: "User no longer exists." });
+        } else {
+          console.log("Document data:", doc.data());
```
Comment (schettn) [inline src/store/actions/userActions.js]: add newline or remove the console log

File: src/store/reducers/authReducer.js [hunk 30]
```diff
@@ -31,6 +31,30 @@ const authReducer = (state = initState, action) => {
         ...state,
         catDetails: null,
       };
+    case "GETZOMBIE_SUCCESS":
+      console.log("Zombies received");
```
Comment (schettn) [inline src/store/reducers/authReducer.js]: add newline or remove the console log

File: src/store/reducers/authReducer.js [hunk 31]
```diff
@@ -31,6 +31,30 @@ const authReducer = (state = initState, action) => {
...
+      return {
+        ...state,
+        zombies: action.zombies,
+      };
+    case "GETZOMBIE_ERROR":
+      console.log("Can not get zombies", action.err);
```
Comment (schettn) [inline src/store/reducers/authReducer.js]: add newline or remove the console log

File: src/store/reducers/authReducer.js [hunk 32]
```diff
@@ -31,6 +31,30 @@ const authReducer = (state = initState, action) => {
...
+      return {
+        ...state,
+        zombies: null,
+      };
+    case "UPGRADE_SUCCESS":
+      console.log("Cat upgraded");
```
Comment (schettn) [inline src/store/reducers/authReducer.js]: add newline or remove the console log

File: src/store/reducers/authReducer.js [hunk 33]
```diff
@@ -31,6 +31,30 @@ const authReducer = (state = initState, action) => {
...
+      return {
+        ...state,
+        upgradeFailed: false,
+      };
+    case "UPGRADE_ERROR":
+      console.log("Can not upgrade cat", action.err);
```
Comment (schettn) [inline src/store/reducers/authReducer.js]: add newline or remove the console log

File: src/store/reducers/authReducer.js [hunk 34]
```diff
@@ -31,6 +31,30 @@ const authReducer = (state = initState, action) => {
...
+      return {
+        ...state,
+        upgradeFailed: true,
+      };
     case "REMOVECAT_ERROR":
       console.log("Can not disable cat", action.err);
```
Comment (schettn) [inline src/store/reducers/authReducer.js]: add newline or remove the console log

File: src/store/reducers/postReducer.js [hunk 35]
```diff
@@ -1,105 +1,105 @@
...
-      case "COMMENT_ERROR":
-        console.log("Comment not created", action.err);
-        return {
-          ...state,
-        };
-      default:
-        return state;
+  switch (action.type) {
+    case "CREATION_SUCCESS":
+      console.log("Creation success");
```
Comment (schettn) [inline src/store/reducers/postReducer.js]: add newline or remove the console log

File: src/store/reducers/postReducer.js [hunk 36]
```diff
@@ -1,105 +1,105 @@
...
+      return {
+        ...state,
+        postError: null,
+      };
+    case "CREATION_ERROR":
+      console.log("Creation error", action.err);
```
Comment (schettn) [inline src/store/reducers/postReducer.js]: add newline or remove the console log

File: src/store/reducers/postReducer.js [hunk 37]
```diff
@@ -1,105 +1,105 @@
...
+      return {
+        ...state,
+        postError: action.err,
+      };
+    case "CREATION_SPAM":
+      console.log("Post is spam", action.err);
```
Comment (schettn) [inline src/store/reducers/postReducer.js]: add newline or remove the console log

File: src/store/reducers/postReducer.js [hunk 38]
```diff
@@ -1,105 +1,105 @@
...
+      return {
+        ...state,
+        postError: action.err,
+      };
+    case "LOAD_SUCCESS":
+      console.log("Loading success");
```
Comment (schettn) [inline src/store/reducers/postReducer.js]: add newline or remove the console log

File: src/store/reducers/postReducer.js [hunk 39]
```diff
@@ -1,105 +1,105 @@
...
+      return {
+        ...state,
+        authError: null,
+        results: action.results,
+        loading: false,
+      };
+    case "LOAD_LOADING":
+      console.log("Loading...");
```
Comment (schettn) [inline src/store/reducers/postReducer.js]: add newline or remove the console log

File: src/store/reducers/postReducer.js [hunk 40]
```diff
@@ -1,105 +1,105 @@
...
+      return {
+        ...state,
+        loading: true,
+      };
+    case "LOAD_ERROR":
+      console.log("Loading error", action.err);
```
Comment (schettn) [inline src/store/reducers/postReducer.js]: add newline or remove the console log

File: src/store/reducers/postReducer.js [hunk 41]
```diff
@@ -1,105 +1,105 @@
...
+      return {
+        ...state,
+        authError: action.err,
+        loading: false,
+      };
+    case "LIKE_SUCCESS":
+      console.log("Like success");
```
Comment (schettn) [inline src/store/reducers/postReducer.js]: Okay you should check the issue above in the whole file.. :)

File: src/store/reducers/userReducer.js [hunk 42]
```diff
@@ -1,15 +1,15 @@
...
 const initState = {
   receivedUser: undefined,
-}
+};
 
 const userReducer = (state = initState, action) => {
-  switch(action.type){
+  switch (action.type) {
     case "USERLOAD_SUCCESS":
       console.log("Get user info success");
```
Comment (schettn) [inline src/store/reducers/userReducer.js]: add newline or remove the console log. Check for this issue in the whole file.

**Review by schettn — CHANGES_REQUESTED** [review]
I would suggest to add a new line.

**Review by pinterid — APPROVED** [review]
Looks good to merge after the changes @schettn requested are implemented.

File: src/components/molecules/modals/AboutUs/index.jsx (hunk 8 above)
Comment (Aichnerc) [reply src/components/molecules/modals/AboutUs/index.jsx]: This is a curried function, which does not make sense in this case. It has been removed and converted to a regular function as you requested.

File: src/components/organisms/ZombieList/index.jsx (hunk 9 above)
Comment (Aichnerc) [reply src/components/organisms/ZombieList/index.jsx]: This entire function was useless in this component and has been removed.

File: src/store/actions/postActions.js (hunk 15 above)
Comment (Aichnerc) [reply src/store/actions/postActions.js]: The entire file has been removed, once it was no longer in use.

**Review by schettn — APPROVED** [review]
LGTM!

**Pull request merged**